packages/tizen-remote/test/unit/tizen-remote.spec.js

```diff
@@ -0,0 +1,269 @@
+import {EventEmitter} from 'node:events';
+import {createSandbox} from 'sinon';
+import unexpected from 'unexpected';
+import unexpectedSinon from 'unexpected-sinon';
+import unexpectedEventEmitter from 'unexpected-eventemitter';
+import rewiremock from 'rewiremock/node';
+
+const expect = unexpected.clone().use(unexpectedSinon).use(unexpectedEventEmitter);
+
+describe('TizenRemote', function () {
+  /**
+   * @type {sinon.SinonSandbox}
+   */
+  let sandbox;
+
+  /** @type {typeof import('../../lib/index').TizenRemote} */
+  let TizenRemote;
+
+  /** @type {typeof import('../../lib/index').WsEvent} */
+  let WsEvent;
+
+  /** @type {typeof import('../../lib/index').Event} */
+  let Event;
+
+  let mocks;
+
+  /** @type {typeof import('../../lib/index').constants} */
+  let constants;
+
+  /** @type {{get: sinon.SinonStub}} */
+  let mockEnv;
+
+  /** @type {InstanceType<MockWebSocket>} */
+  let mockWs;
+
+  /** @type {new (url: string, opts?: import('ws').ClientOptions) => EventEmitter & {close: () => void, send: (msg: string, done: ((err?: Error) => void)) => void}} */
+  let MockWebSocket;
+
+  beforeEach(function () {
+    sandbox = createSandbox();
+
+    mockEnv = {get: sandbox.stub()};
+
+    MockWebSocket = class MockWebSocket extends EventEmitter {
+      /**
+       * @param {string} url
+       * @param {import('ws').ClientOptions} [opts]
+       */
+      constructor(url, opts) {
+        super();
+        this.url = url;
+        this.opts = opts;
+        this.readyState = MockWebSocket.CONNECTING;
+
+        this.close = sandbox.stub().callsFake(() => {
+          // this happens synchronously
+          this.readyState = MockWebSocket.CLOSING;
+          setTimeout(() => {
+            this.readyState = MockWebSocket.CLOSED;
+            this.emit(WsEvent.CLOSE);
+          });
+        });
+
+        this.send = sandbox.stub().callsFake((msg, done) => {
+          setTimeout(() => {
+            done();
+            setTimeout(() => {
+              this.emit(
+                WsEvent.MESSAGE,
+                JSON.stringify({event: constants.TOKEN_EVENT, data: {token: 'token'}})
+              );
+            });
+          });
+        });
+
+        this.init();
+
+        mockWs = this;
+      }
+
+      init() {
```

Comment:
the `WebSocket` client unfortunately does a _highly questionable thing_, and the thing it does is adds tasks to the even loop _in its constructor_ and forces you to listen for its `open` even _or_ pass a rather obscure node-style callback into its options object.  this means that to reasonably mimic the class, the logic for connection has to live in the constructor as well.  again, _this sucks_, so I added this `init` method to be able to easily swap out the behavior.

Reply:
FWIW, I'm finding that using private methods and fields feels really good because it prevents writing too many unit tests against stuff you shouldn't.  that isn't to say I'm "done" with the unit tests, but I'm done for now.